cdci_data_analysis/flask_app/dispatcher_query.py

```diff
@@ -1490,15 +1497,29 @@ def set_instrument(self, instrument_name, roles, email):
         if instrument_name == 'mock':
             new_instrument = 'mock'
         else:
-            for instrument_factory in importer.instrument_factory_list:
-                instrument = instrument_factory()
-                if instrument.name == instrument_name:
-                    if instrument.instrumet_query.check_instrument_access(roles, email):
-                        new_instrument = instrument  # multiple assignment? TODO
+            for instrument_factory in importer.instrument_factory_iter:
+                _instrument = None
+                if hasattr(instrument_factory, 'instr_name'):
+                    instr_name = instrument_factory.instr_name
+                else:
+                    _instrument = instrument_factory()
+                    instr_name = _instrument.name
+                
+                if instr_name == instrument_name:
+                    if _instrument is None and hasattr(instrument_factory, 'instrument_query'):
+                        instr_query = instrument_factory.instrument_query
+                    else:
+                        _instrument = instrument_factory()
```

Comment:
should this go one line up ?

Reply:
I this I am not understand why the `_instrument = instrument_factory()` is executed in case `_instrument is not None`

Reply:
Yep, something wrong here, thanks!

Reply:
Fixed. Variants are:

- Old plugins: factory functuin has no `instr_name`, `instrument_query` attributes. Then Instrument is created in the first else clause, regardless of that name is the one we request, getting `name` and `check_instrument_access` from the instance, as before.
- Both `instr_name` and `instrument_query` are attached to factory function. Instrument instance will be created only if one requests it and has access to it. Otherwise only list of names `known_instruments` is needed.
- Only `instr_name` factory attribute is available. Then the instance will be created to check access if one requests this particular instrument. This behaviour is not currently needed, just in case for the future (e.g. migrating plugins with non-trivial InstrumentQuery).